### PR TITLE
Use stack instead of recursion in MCScanner>>next

### PR DESCRIPTION
**Pull request by jgfoster**
This avoids a stack overflow when loading Metacello into GemStone.
I've also added a number of comments (which accounts for some of the extra length).

### CODE REVIEW

File: src/System-Hashing/MD5NonPrimitive.class.st [hunk 1]
```diff
@@ -25,7 +25,7 @@ Class {
 
 { #category : #'class initialization' }
 MD5NonPrimitive class >> initialize [
-	"MD5 initialize"
+	"MD5NonPrimitive initialize"
 	"Obscure fact: those magic hex numbers that are hard to type in correctly are
```
Comment (jgfoster) [inline src/System-Hashing/MD5NonPrimitive.class.st:29]: This isn't related to the other edit, but was just a comment I stumbled across that seems wrong.

File: src/Monticello/MCScanner.class.st [hunk 2]
```diff
@@ -27,14 +27,32 @@ MCScanner class >> scanTokens: aString [
...
-	c = $' ifTrue: [^ self nextString].
-	c = $( ifTrue: [^ self nextArray].
-	c isAlphaNumeric ifTrue: [^ self nextSymbol].
-	self error: 'Unknown token type'.	
+	"This would be slightly simpler using recursion but it risks a stack overflow 
+	 for some packages on some platforms, so we implement it using a local stack. 
+	 The performance is essentially equivalent (<5% difference)."
+
+	| stack |
+	stack := OrderedCollection with: OrderedCollection new.
```
Comment (gcotelli) [inline src/Monticello/MCScanner.class.st]: Why not using `Stack` as the stack?

Comment (jgfoster) [reply src/Monticello/MCScanner.class.st]: Because I'm used to GemStone that doesn't have it. I've submitted an update. Thanks!

**Comment by Ducasse** [comment]
James the following tests are failing and they look related to your changes.
unix-32 / Tests-unix-32 / test4 – Unix32.Monticello.Tests.Base.MCScannerTest
<1s
unix-32 / Tests-unix-32 / test5 – Unix32.Monticello.Tests.Base.MCScannerTest

**Comment by jgfoster** [comment]
Those tests certainly would seem related to my changes, but I'm not able to see the failures. It seems to me that as of commit 4d27c0e the Unix32 tests passed.

Screen Shot 2019-08-18 at 10 44 26 AM

**Comment by jgfoster** [comment]
I'm able to see the failures now. I'll investigate.

**Comment by jgfoster** [comment]
MCScannerTest now passes.

**Review by MarcusDenker — APPROVED** [review]

**Pull request merged**